### PR TITLE
ptc-1765 | Adding '/forex/rates' api spec for UPAPI NAS

### PR DESCRIPTION
**Pull request by chaz-doyle-cko**
# Description of changes in PR

[PTC-1765](https://checkout.atlassian.net/browse/PTC-1765)

[//]: # 'Adding /forex/rates endpoint for new public-fx-rates api'

## Checklist

- [x] Added a changelog entry to the `changelog.md` file in either `abc_spec` or `nas_spec`
- [x] Contacted the Tech Docs team about any corresponding guides that need to be updated for www.checkout.com/docs

**Contributing options**: Look at [our documentation](https://checkout.atlassian.net/wiki/spaces/PD/pages/2169506663/API+ref+publication+process) for options to contribute to the API reference.

[PTC-1765]: https://checkout.atlassian.net/browse/PTC-1765?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: nas_spec/components/schemas/Forex/ForexRatesRatesObject.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+type: object
+properties:
+  exchange_rate:
+    description: The FX rate with Checkout margin applied. Up to 8 decimal places provided in response.
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/components/schemas/Forex/ForexRatesRatesObject.yaml:4]:
```suggestion
    description: The FX rate with Checkout.com margin applied. Up to eight decimal places provided in response.
```

File: nas_spec/components/schemas/Forex/ForexRatesResponse.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  source:
+    $ref: '#/components/schemas/ForexSource'
+  rates:
+    description: List of FX rates with Checkout margin applied
+    schema:
+      type: array
+      items:
+        $ref: '#/components/schemas/ForexRatesRatesObject'
+  invalid_currency_pairs:
+    description: List of currency pairs that Checkout was unable to provide rates for
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/components/schemas/Forex/ForexRatesResponse.yaml:13]:
```suggestion
    description: List of currency pairs that Checkout.com was unable to provide rates for
```

File: nas_spec/components/schemas/Forex/properties/ForexProduct.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+type: string
+enum: 
+  - card_payouts
+description: The Checkout product to return rates for
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/components/schemas/Forex/properties/ForexProduct.yaml:4]:
```suggestion
description: The Checkout.com product to return rates for
```

File: nas_spec/paths/forex@rates.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
+get:
+  tags:
+    - Forex
+  security:
+    - OAuth:
+        - fx
+  summary: Get FX rates
+  operationId: getFxRates
+  description: |
+    Request FX rates for a specific Checkout product and rate source.
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/paths/forex@rates.yaml:10]:
```suggestion
    Request FX rates for a specific Checkout.com product and rate source.
```

File: nas_spec/paths/forex@rates.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        application/json:
+          schema:
+            $ref: '#/components/schemas/ForexRatesResponse'
+      headers:
+        Cko-Request-Id:
+          $ref: '#/components/headers/Cko-Request-Id'
+    '401':
+      description: Unauthorized
+    '429':
+      description: Too many requests or duplicate request detected
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/paths/forex@rates.yaml:47]:
```suggestion
      description: Too many requests or duplicate requests detected
```

File: nas_spec/components/schemas/Forex/ForexRatesResponse.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
+type: object
+properties:
+  product:
+    $ref: '#/components/schemas/ForexProduct'
+  source:
+    $ref: '#/components/schemas/ForexSource'
+  rates:
+    description: List of FX rates with Checkout margin applied
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/components/schemas/Forex/ForexRatesResponse.yaml:8]:
```suggestion
    description: List of FX rates with Checkout.com margin applied
```

File: nas_spec/changelog.md [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 | Date       | Description of change                                                                                                 |
 |------------|-----------------------------------------------------------------------------------------------------------------------|
+| 2023/03/16 | Added `/forex/rates` UPAPI endpoint to NAS spec                                                                       |
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/changelog.md:5]: Please ignore this comment if I've suggest wrong explanation for acronym. :)

```suggestion
| 2023/03/16 | Added `/forex/rates` Unified Payments API (UPAPI) endpoint to NAS spec                                                                       |
```

File: nas_spec/components/schemas/Forex/properties/ForexProduct.yaml (hunk 3 above)
Comment (chaz-doyle-cko) [reply nas_spec/components/schemas/Forex/properties/ForexProduct.yaml]: Amended

File: nas_spec/paths/forex@rates.yaml (hunk 4 above)
Comment (chaz-doyle-cko) [reply nas_spec/paths/forex@rates.yaml]: Amended

File: nas_spec/components/schemas/Forex/ForexRatesResponse.yaml (hunk 6 above)
Comment (chaz-doyle-cko) [reply nas_spec/components/schemas/Forex/ForexRatesResponse.yaml]: Amended

File: nas_spec/changelog.md (hunk 7 above)
Comment (chaz-doyle-cko) [reply nas_spec/changelog.md]: Yep, you're completely correct here. 😄

File: nas_spec/components/schemas/Forex/ForexRate.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+type: object
+properties:
+  exchange_rate:
+    description: The FX rate with Checkout.com margin applied. Up to 8 decimal places provided in response.
```
Comment (eoin-shaughnessy-cko) [inline nas_spec/components/schemas/Forex/ForexRate.yaml:4]: Suggestion to align with [brand guidelines](https://brand.checkout.com/document/1#/writing-for-cko/numbers/figures):

```suggestion
    description: The FX rate with Checkout.com margin applied. Up to eight decimal places provided in response.
```

File: nas_spec/paths/forex@rates.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        description: Comma-separated list of currency pairs
+        example: 'GBPEUR,USDNOK,JPNCAD'
+    - name: processing_channel_id
+      in: query
+      required: true
+      schema:
+        $ref: '#/components/schemas/ForexProcessingChannelId'
+  responses:
+    '200':
+      description: Rates retreived successfully
```
Comment (philip-wood-cko) [inline nas_spec/paths/forex@rates.yaml:36]: spelling - `retrieved`

**Review by eoin-shaughnessy-cko — APPROVED** [review]

**Review by eoin-shaughnessy-cko — DISMISSED** [review]

**Pull request merged**